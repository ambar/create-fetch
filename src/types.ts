export type Fetch = typeof fetch
export type FetchReturn = ReturnType<Fetch>

/**
 * Enhance fetch
 *
 * **Example**
 *
 * ```ts
 * // with extra arguments
 * type FooInit = {foo?: boolean}
 * const foo = (): FetchEnhancer<FooInit> => (fetch) => (url, {foo, ...options}) => {
 *   if (foo) {
 *     // ...
 *   }
 *   return fetch(url, options)
 * }
 *
 * // no extra arguments
 * const bar = (): FetchEnhancer => (fetch) => (url, options) => {
 *   return fetch(url, options)
 * }
 *
 * const myFetch = foo()(bar()(fetch))
 * ```
 */
// eslint-disable-next-line @typescript-eslint/ban-types
export type FetchEnhancer<Extra = {}> = <F extends Fetch>(
  fetch: F
) => (url: Parameters<F>[0], options?: Parameters<F>[1] & Extra) => FetchReturn

type ExtractGeneric<T> = T extends FetchEnhancer<infer X> ? X : never
type SpreadGeneric<T extends readonly [...any]> = T extends readonly [
  infer L,
  ...infer R
]
  ? ExtractGeneric<L> & SpreadGeneric<R>
  : unknown

/**
 * Compose fetch
 *
 * **Example**
 *
 * ```ts
 * const myFetch = (myCompose as FetchCompose)(query, bodify)(fetch)
 * ```
 */
export type FetchCompose = <FEs extends FetchEnhancer[]>(
  ...args: FEs
) => <F extends Fetch>(
  fetch: F
) => (
  url: Parameters<F>[0],
  options?: Parameters<F>[1] & SpreadGeneric<FEs>
) => FetchReturn

export type FetchCreate = <
  F extends Fetch,
  FEs extends readonly FetchEnhancer[]
>(
  fetch: F,
  args: FEs
) => (
  url: Parameters<F>[0],
  options?: Parameters<F>[1] & SpreadGeneric<FEs>
) => FetchReturn
