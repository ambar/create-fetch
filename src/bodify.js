import toHeaders from './utils/toHeaders'
import toObject from './utils/toObject'

const isPlainObject = obj =>
  obj !== null &&
  typeof obj === 'object' &&
  Object.getPrototypeOf(obj) === Object.prototype

/**
 * Stringify body
 */
const bodify = () => fetch => (url, {body, ...options} = {}) => {
  let headers = toHeaders(options.headers)
  const contentType = 'content-type'
  if (body instanceof URLSearchParams) {
    // patches old browsers
    if (!headers.get(contentType)) {
      headers.set(contentType, 'application/x-www-form-urlencoded')
    }
  } else if (isPlainObject(body) || Array.isArray(body)) {
    body = JSON.stringify(body)
    if (!headers.get(contentType)) {
      headers.set(contentType, 'application/json')
    }
  }

  return fetch(url, {
    ...options,
    body,
    headers: toObject(headers),
  })
}

export default bodify
