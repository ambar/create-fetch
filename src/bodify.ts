import toHeaders from './utils/toHeaders'
import toObject from './utils/toObject'
import isPlainObject from './utils/isPlainObject'
import {FetchEnhancer} from './types'

/* TODO: override body type
type JSONValue = string | number | boolean | JSONObject | JSONArray

interface JSONObject {
  [x: string]: JSONValue
}

interface JSONArray extends Array<JSONValue> {}

type BodifyInit = {
  body: JSONObject | JSONArray | BodyInit
}
*/

/**
 * Stringify body
 */
const bodify =
  (): FetchEnhancer =>
  (fetch) =>
  (url, {body, ...options} = {}) => {
    let headers = toHeaders(options.headers)
    const contentType = 'content-type'
    if (body instanceof URLSearchParams) {
      // patches old browsers
      if (!headers.get(contentType)) {
        headers.set(
          contentType,
          'application/x-www-form-urlencoded;charset=UTF-8'
        )
      }
    } else if (isPlainObject(body) || Array.isArray(body)) {
      body = JSON.stringify(body)
      if (!headers.get(contentType)) {
        headers.set(contentType, 'application/json')
      }
    }

    return fetch(url, {
      ...options,
      body,
      headers: toObject(headers),
    })
  }

export default bodify
