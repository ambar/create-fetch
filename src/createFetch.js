const compose = (...fns) => arg => fns.reduceRight((acc, f) => f(acc), arg)

const createFetch = (fetch, middlewares = []) => compose(...middlewares)(fetch)

export default createFetch
