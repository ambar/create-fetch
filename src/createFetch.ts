import {FetchEnhancer, FetchCompose, FetchCreate} from './types'

const compose =
  (...fns) =>
  (arg) =>
    fns.reduceRight((acc, f) => f(acc), arg)

export const composeFetch: FetchCompose = compose
export const createFetch: FetchCreate = (fetch, enhancers: FetchEnhancer[] = []) => compose(...enhancers)(fetch)
export default createFetch
