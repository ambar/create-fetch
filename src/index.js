export {default} from './createFetch'
export {default as createFetch} from './createFetch'
export {default as defaults} from './defaults'
export {default as headers} from './headers'
export {default as query} from './query'
export {default as bodify} from './bodify'
export {default as xsrf} from './xsrf'
