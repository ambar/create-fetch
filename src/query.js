/**
 * Stringify query string
 */
const query = () => fetch => (url, {query, ...options} = {}) => {
  if (query !== null && typeof query === 'object') {
    const params = (query instanceof URLSearchParams
      ? query
      : new URLSearchParams(query)
    ).toString()
    if (params) {
      const index = url.lastIndexOf('?')
      url += (index === -1 ? '?' : index === url.length - 1 ? '' : '&') + params
    }
  }

  return fetch(url, options)
}

export default query
