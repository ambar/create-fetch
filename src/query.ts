import getEntries from './utils/getEntries'
import {FetchEnhancer} from './types'

// preserve legacy querystring/qs behavior, set null or undefined to empty
const mapNilToEmpty = (obj: Record<string, any>) =>
  getEntries(obj).map(([k, v]) => [k, (v as string) ?? ''])

type QueryInit = {
  query?: string[][] | Record<string, any> | URLSearchParams
}

/**
 * Stringify query string
 */
const query =
  (): FetchEnhancer<QueryInit> =>
  (fetch) =>
  (url, {query, ...options} = {}) => {
    if (query !== null && typeof query === 'object') {
      const params = (
        query instanceof URLSearchParams
          ? query
          : new URLSearchParams(mapNilToEmpty(query))
      ).toString()
      if (params) {
        url = String(url)
        const index = url.lastIndexOf('?')
        url +=
          (index === -1 ? '?' : index === url.length - 1 ? '' : '&') + params
      }
    }

    return fetch(url, options)
  }

export default query
