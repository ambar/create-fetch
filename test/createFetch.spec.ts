/* eslint-disable @typescript-eslint/no-unsafe-assignment */
import createFetch from '../src/createFetch'
import query from '../src/query'
import bodify from '../src/bodify'
import {FetchEnhancer} from '../src/types'

describe('createFetch', () => {
  const fetch = jest.fn(
    (async (...args) => args) as unknown as typeof globalThis.fetch
  )

  it('should create fetch without middlewares', async () => {
    const myFetch = createFetch(fetch, [])
    expect(await myFetch('/api')).toMatchSnapshot()
    expect(
      await myFetch('/api', {headers: {'x-version': '1'}})
    ).toMatchSnapshot()
  })

  it('should create fetch with middlewares', async () => {
    const presets = [query(), bodify()] as const
    const myFetch = createFetch(fetch, presets)

    // default
    expect(await myFetch('/api')).toMatchSnapshot()
    // query string
    expect(await myFetch('/api', {query: {filter: 'name'}})).toMatchSnapshot()
    // json
    expect(
      await myFetch('/api', {method: 'POST', body: {name: 'JoJo'} as any})
    ).toMatchSnapshot()
    // form
    expect(
      await myFetch('/api', {
        method: 'POST',
        query: {filter: 'name'},
        body: new URLSearchParams({name: 'JoJo'}),
      })
    ).toMatchSnapshot()
  })

  it('should create fetch with compose', async () => {
    const next =
      (
        onNext: (...args: Parameters<typeof globalThis.fetch>) => void
      ): FetchEnhancer =>
      (fetch) =>
      (url, options) => {
        onNext(url, options)
        return fetch(url, options)
      }

    const onNext = jest.fn()
    const myFetch = createFetch(fetch, [query(), next(onNext)] as const)
    await myFetch('/api', {method: 'POST', query: {foo: 1}})
    expect(onNext).toBeCalledWith('/api?foo=1', {method: 'POST'})
  })
})
