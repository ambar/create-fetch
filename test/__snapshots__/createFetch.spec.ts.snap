// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`createFetch should create fetch with middlewares 1`] = `
[
  "/api",
  {
    "body": undefined,
    "headers": {},
  },
]
`;

exports[`createFetch should create fetch with middlewares 2`] = `
[
  "/api?filter=name",
  {
    "body": undefined,
    "headers": {},
  },
]
`;

exports[`createFetch should create fetch with middlewares 3`] = `
[
  "/api",
  {
    "body": "{"name":"JoJo"}",
    "headers": {
      "content-type": "application/json",
    },
    "method": "POST",
  },
]
`;

exports[`createFetch should create fetch with middlewares 4`] = `
[
  "/api?filter=name",
  {
    "body": URLSearchParams: "name=JoJo",
    "headers": {
      "content-type": "application/x-www-form-urlencoded;charset=UTF-8",
    },
    "method": "POST",
  },
]
`;

exports[`createFetch should create fetch without middlewares 1`] = `
[
  "/api",
]
`;

exports[`createFetch should create fetch without middlewares 2`] = `
[
  "/api",
  {
    "headers": {
      "x-version": "1",
    },
  },
]
`;
